### PR TITLE
Dagandreas-5-epic-user-profile-page

### PR DESCRIPTION
**Pull request by DagAndreas**

### CODE REVIEW

**Review by JosteinRuen — DISMISSED** [review]
Amazing work Dag!

**Review by johnfa1508 — APPROVED** [review]
Good job Dag!

File: src/pages/profile/index.js [hunk 1]
```diff
@@ -1,7 +1,264 @@
...
+  const { id } = useParams();
+  const [profileUser, setProfileUser] = useState(null);
+  const [loggedInUser, setLoggedInUser] = useState(null);
+  const { token } = useAuth();
+  const [canEdit, setCanEdit] = useState(false);
+  const [formData, setFormData] = useState(null);
+  const [prevProfileUserId, setPrevProfileUserId] = useState(null); // New state variable
+
+  useEffect(() => {
+    const fetchUserData = async (profileId, setter) => {
```
Comment (dogezen) [inline src/pages/profile/index.js:22]: this function can be defined outside of useEffect, like on line 20

File: src/pages/profile/index.js [hunk 2]
```diff
@@ -1,7 +1,264 @@
...
+      try {
+        const response = await get(`users/${profileId}`);
+        setter(response.data);
+      } catch (error) {
+        console.error(`Error fetching user data for ID ${profileId}:`, error);
+      }
+    };
+
+    if (token) {
+      const loggedInId = jwtDecode(token).userId;
```
Comment (dogezen) [inline src/pages/profile/index.js:32]: you should be able to get this from the auth context -> the logged in user is available from the auth context, so there should not be a need to re-fetch the data?

File: src/pages/profile/index.js [hunk 3]
```diff
@@ -1,7 +1,264 @@
...
+
+    Object.keys(formData).forEach((key) => {
+      if (!excludedFields.includes(key) && formData[key] !== profileUser[key]) {
+        changeData[key] = formData[key];
+      }
+    });
+
+    if (Object.keys(changeData).length > 0) {
+      try {
+        const response = await patch(`users/${profileUser.id}`, changeData);
```
Comment (dogezen) [inline src/pages/profile/index.js:74]: why are you not using the pattern we have in the repo to send API requests of writing the fetch directly in the service/apiClient.js?

https://github.com/boolean-uk/team-dev-client-ex-2410-team1/blob/main/src/service/apiClient.js

File: src/pages/profile/index.js [hunk 4]
```diff
@@ -1,7 +1,264 @@
...
+    }
+  }, [id, token]);
+
+  useEffect(() => {
+    if (profileUser && loggedInUser) {
+      setCanEdit(profileUser.id === loggedInUser.id || loggedInUser.role === 'teacher');
+    }
+  }, [profileUser, loggedInUser]);
+
+  useEffect(() => {
```
Comment (dogezen) [inline src/pages/profile/index.js:44]: good

File: src/pages/profile/index.js [hunk 5]
```diff
@@ -1,7 +1,264 @@
...
+    };
+
+    if (token) {
+      const loggedInId = jwtDecode(token).userId;
+      fetchUserData(id, setProfileUser); // Fetch profile user data
+      fetchUserData(loggedInId, setLoggedInUser); // Fetch logged-in user data
+    }
+  }, [id, token]);
+
+  useEffect(() => {
```
Comment (dogezen) [inline src/pages/profile/index.js:38]: good

File: src/service/apiClient.js [hunk 6]
```diff
@@ -52,4 +52,4 @@ async function request(method, endpoint, data, auth = true) {
   return response.json();
 }
 
-export { login, getPosts, register, createProfile };
+export { login, getPosts, register, createProfile, request, get, post, patch };
```
Comment (dogezen) [inline src/service/apiClient.js:55]: don't export request, get, post, patch -> instead define a new function in this module called `updateUserProfile` that takes in the data to update, the user ID and sends the request for you and returns the response payload, see the `createprofile` for example.

File: src/context/auth.js [hunk 7]
```diff
@@ -48,18 +50,15 @@ const AuthProvider = ({ children }) => {
     }
 
     localStorage.setItem('token', res.data.token);
-    localStorage.setItem('loggedInUser', JSON.stringify(res.data.user));
-
     setToken(res.data.token);
     setLoggedInUser(res.data.user);
-
+    localStorage.removeItem('redirectPath');
```
Comment (dogezen) [inline src/context/auth.js:55]: double check with Team lead that this OK to add

File: src/context/auth.js [hunk 8]
```diff
@@ -48,18 +50,15 @@ const AuthProvider = ({ children }) => {
...
     navigate('/');
   };
 
   const handleLogout = () => {
     localStorage.removeItem('token');
-    localStorage.removeItem('loggedInUser');
-
+    localStorage.removeItem('redirectPath');
```
Comment (dogezen) [inline src/context/auth.js:61]: double check with Team lead that this OK to add

**Review by dogezen — DISMISSED** [review]
Please update the code to:

1) load logged in user data from auth context, not send a new fetch request
2) the get user profile / and update user profile API requests, should go in services/apiClient.js and be exported as functions, don't use post/get functions directly
3) with the auth context changes - please double check with team lead, i know there was a problem with navigation so your change might break navigation again

File: src/service/apiClient.js (hunk 6 above)
Comment (DagAndreas) [reply src/service/apiClient.js:55]: Do i have to?  Others on the team also rely on the same exports. They just haven't pr'd it. 
I do understand the benefits of declaring what requests are made in the same file though.

File: src/pages/profile/index.js (hunk 2 above)
Comment (DagAndreas) [reply src/pages/profile/index.js:32]: The logged in user gets set to null when i reload the page, which is why i rely on fetching every time.

**Comment by DagAndreas** [comment]
Merged with main (johns update)

refactored out the get and patch functions

still fetching on logged in due to errors with loggedincontext

**Review by JosteinRuen — APPROVED** [review]
Awesome

**Review by johnfa1508 — APPROVED** [review]

File: src/components/header/index.js [hunk 9]
```diff
@@ -25,20 +37,23 @@ const Header = () => {
...
+        <p>{initials}</p>
       </div>
 
       {isMenuVisible && (
         <div className="user-panel">
           <Card>
             <section className="post-details">
               <div className="profile-icon">
-                <p>AJ</p>
+                {/* was AJ */}
```
Comment (allendewilzin) [inline src/components/header/index.js:48]: remove this line

File: src/pages/profile/index.js [hunk 10]
```diff
@@ -1,7 +1,256 @@
...
+import useAuth from '../../hooks/useAuth';
+import Form from '../../components/form';
+import TextInput from '../../components/form/textInput';
+
+import './styles.css';
+
 const ProfilePage = () => {
+  const { id } = useParams();
+  const [profileUser, setProfileUser] = useState(null);
+  // const [loggedInUser, setLoggedInUser] = useState(null);
```
Comment (allendewilzin) [inline src/pages/profile/index.js:14]: unused lines please remove

**Review by allendewilzin — CHANGES_REQUESTED** [review]

**Review by allendewilzin — DISMISSED** [review]

**Comment by DagAndreas** [comment]
Unused line removed

**Review by JosteinRuen — APPROVED** [review]
👍

**Review by johnfa1508 — APPROVED** [review]

**Pull request merged**